### ISSUE TITLE
2 issues with naive bayes - prediction is same every time; and memory exception if more than couple hundred records in training data

### ISSUE THREAD

**Issue by groffg** [sethjuarez/numl#40]
I must be doing something horribly wrong. I'm trying to use naive bayes to categorize some data based on input via a spreadsheet. The code runs, but it seems to give me the same category regardless of my input. I've appended some sample code below (gloss over the helper call that reads Excel; that just takes the sheet & converts to a .net dataset). Also, I'm limiting the training data to the first 100 rows. Any more than that and the program gets an out of memory exception!

So I have two questions:
* how to get a meaningful prediction/classification?
* how to load much larger training dataset without memory overflow?

```c#
	[Serializable]
	public class PeerCategory
	{
		[Feature]
		public string PeerCategoryDesc { get; set; }
		[Label]
		public string CAPeerCategory { get; set; }
	}

	[TestFixture]
	//[Ignore]
	public class BayesTest
	{

		[Test]
		public void TestExcel()
		{
			string pathToFile = @"C:\data\Training Data.xlsx";
			using (var reader = new FileReaderExcel(pathToFile))
			{
				int index = reader.GetSheetIndexByName("V6");
				var dt = reader.GetDataTableFromExcelSheet(index, true);
				var peerCatList = (
					from DataRow row in dt.Rows
					select new PeerCategory()
					{
						PeerCategoryDesc = row.Field<string>(0), 
						CAPeerCategory = row.Field<string>(1)
					}).Take(100).ToList();


				var width = peerCatList.Select(t => t.CAPeerCategory).Distinct().Count();

				IGenerator generator = new NaiveBayesGenerator(width);
				generator.Descriptor = Descriptor.Create<PeerCategory>();
				LearningModel learned = Learner.Learn(peerCatList, 0.8, 1000, generator);

				IModel model = learned.Model;
				double accuracy = learned.Accuracy;

			
				var value1 = model.Predict(GetItem("paint"));
				var value2 = model.Predict(GetItem("flower"));
				var value3 = model.Predict(GetItem("sofa"));
				var value4 = model.Predict(GetItem("desk"));
				var value5 = model.Predict(GetItem("bones"));
                                // value1 thru 5 will be the same category over and over again!!

			}
	
		}

		public PeerCategory GetItem(string desc)
		{
			var item = new PeerCategory()
			{
				PeerCategoryDesc = desc,
				CAPeerCategory = string.Empty
			};
			return item;
		}
```

**Comment by bdschrisk** [comment]
To try and resolve the issue can you try reducing the maximum learning iterations down from 1000 to 1 and work up from there.
I.e. change the Learner.Learn line to:
   LearningModel learned = Learner.Learn(peerCatList, 0.8, 1, generator);

Also try closing the reader (see using block) before running the Learner method to reduce the memory footprint.

Are you able to submit a sample project so that we can take a further look?

**Comment by groffg** [comment]
bdschrisk - thanks for the quick reply. Reducing the iterations from 1000 helped; I also closed/disposed the reader like you suggested. I'm now able to load a larger # of rows in the training data spreadsheet, which is great news.

Now the issue is that the predictions aren't right. I feel like I'm missing something in the code. The training data I'm using came from another team and should be representative of the actual data to be categorized, but it seems that I'm not getting the right result back every time, even when the keywords in the test clearly match those in the desired category.

**Comment by sethjuarez** [comment]
Looks like you are using a StringFeature. I would look at removing stopwords.

**Comment by bill-mybiz** [comment]
Also, could you give just a couple actual training examples that you are feeding into it?

**Comment by groffg** [comment]
Seth - removing all stopwords, or some? How would I go about removing these stopwords?

Bill - sure, some examples are:

Field | Category
------------ | -------------
925 Sterling Argent MICRO PAVE Bracelet - BROOKLYN 20mm | Apparel - Jewelry - Fashion
ELECTRIC 40FIVE SUNGLASSES Red Sea Frame / Melanin Grey Lens EE12347901 | Apparel - Clothing - Men

What I'm trying to do is categorize.. i.e., provide the "Category" when one is not present.

**Comment by sethjuarez** [comment]
Instead of using the Feature attribute use the StringFeature attribute. You can pass in a file that has a list of stopwords the system will remove. You should also add things like Ave, St, etc in your particular case. The system by default uses the bag of words model and spaces for word separation. Hope this helps!

**Comment by groffg** [comment]
I think my primary issue was in my training data, though I've gone ahead & changed to StringFeature. Now I'm experimenting with larger sets of training data & also with how large of a repeat value I can safely use to avoid out of memory exceptions. So far I think I'm on the right track as far as getting back higher quality predicted values.

Thanks a bunch for everyone's help. I'll leave this thread open for another couple days in case anything else comes up or if anyone has other feedback & suggestions (which are always welcome).

**Comment by sethjuarez** [comment]
Great!! Would love to see if you get good results. There's a fair amount of witch-doctory to this stuff in my opinion.

**Comment by bdschrisk** [comment]
@groffg Depending on how large your training set is, Logistic Regression will outperform Naïve Bayes on larger training sets and it should also be more forgiving on feature selection.

**Comment by groffg** [comment]
@bdschrisk  - thanks! I think a "reasonable" or "average" size of our training data will be a few thousand records (let's say 2-5K), though I'd like to know that it can be greater (perhaps tens of thousands) if need be. As a fun test, I tried with 50K records and it worked! It consumed enough memory that I ran the nunit test as 64-bit rather than x86. This might be overkill, but it was a worthwhile test and I'm glad it worked out.

I'll keep the logistic regression idea in mind though. I'm open to trying out other algorithms and I'll keep everyone posted on which ones seem to work best for my use-case.
